### PR TITLE
[SPARK-2165][YARN]add support for setting maxAppAttempts in the ApplicationSubmissionContext

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
...xt

https://issues.apache.org/jira/browse/SPARK-2165

I still have 2 questions:
- If this config is not set, we should use yarn's corresponding value or a default value(like 2) on spark side?
- Is the config name best? Or "spark.yarn.am.maxAttempts"?

### CODE REVIEW

**Comment by sryza** [comment]
My opinion is that this is more of a general app property than an AM property, so I'd go for `spark.yarn.maxAppAttempts` instead.  That also avoids confusion with the fact that elsewhere we've claimed the `spark.yarn.am.*` namespace for the yarn-client AM.

**Comment by WangTaoTheTonic** [comment]
@sryza Thanks. That makes sense.
@tgravescs How do you think?

**Comment by tgravescs** [comment]
I would prefer to see it called spark.yarn.maxAppAttempts as well.

File: docs/running-on-yarn.md [hunk 1]
```diff
@@ -149,6 +149,14 @@ Most of the configs are the same for Spark on YARN as for other deployment modes
   In cluster mode, use spark.driver.extraJavaOptions instead.
   </td>
 </tr>
+<tr>
+  <td><code>spark.yarn.amMaxAttempts</code></td>
+  <td>yarn.resourcemanager.am.max-attempts in YARN</td>
+  <td>
+  The maximum number of ApplicationMaster attempts.
+  It should not be larger than the global number set by resourcemanager. Otherwise, it will be override.
```
Comment (tgravescs) [inline docs/running-on-yarn.md]: Perhaps rephrase slightly:

The maximum number of attempts that will be made to submit the application. It should be no larger than the global number of max attempts in the YARN configuration.

**Comment by WangTaoTheTonic** [comment]
@tgravescs Rephrase done.
@sryza Perhaps we should keep the name `spark.yarn.amMaxAttempts` to make it consistent with `yarn.resourcemanager.am.max-attempts` in YARN and `mapreduce.am.max-attempts` in MR.

**Comment by WangTaoTheTonic** [comment]
I changed the name to "spark.yarn.maxAppAttempts", though I think `spark.yarn.amMaxAttempts` is more consistent with `yarn.resourcemanager.am.max-attempts` in YARN and `mapreduce.am.max-attempts` in MR.

**Comment by tgravescs** [comment]
Thanks @WangTaoTheTonic  changes look good.

**Pull request closed without merging**